pkg/image/builder/image_builder.go

```diff
@@ -80,85 +81,196 @@ func (b *ImageBuilder) Run(ctx context.Context, targetRef string, layoutPath lay
 	if err != nil {
 		return err
 	}
-	idxManifest, err := idx.IndexManifest()
+	// make a copy of the original manifest for later
+	originalIdxManifest, err := idx.IndexManifest()
 	if err != nil {
 		return err
 	}
+	originalIdxManifest = originalIdxManifest.DeepCopy()
 
-	for _, manifest := range idxManifest.Manifests {
-		switch manifest.MediaType {
-		case types.DockerManifestSchema2:
-			v2format = true
-		case types.OCIManifestSchema1:
-			v2format = false
-		default:
-			return fmt.Errorf("image %q: unsupported manifest format %q", targetRef, manifest.MediaType)
-		}
-
-		img, err := layoutPath.Image(manifest.Digest)
+	/*
+		processImageIndex is a recursive function that allows for traversal of the hierarchy of
+		parent/child indexes that can exist for a multi arch image. There's always
+		at least one index at the root since this is an OCI layout that we're dealing with.
+		In theory there can be "infinite levels" of "index indirection" for multi arch images, but typically
+		its only two levels deep (i.e. index.json itself which is level one, and the manifest list
+		defined in the blobs directory, which is level two).
+	*/
+	var processImageIndex func(idx v1.ImageIndex) (v1.ImageIndex, error)
+	processImageIndex = func(idx v1.ImageIndex) (v1.ImageIndex, error) {
+		var resultIdx v1.ImageIndex
+		resultIdx = idx
+		idxManifest, err := idx.IndexManifest()
 		if err != nil {
-			return err
+			return nil, err
 		}
+		for _, manifest := range idxManifest.Manifests {
+			currentHash := *manifest.Digest.DeepCopy()
+			switch manifest.MediaType {
+			case types.DockerManifestList, types.OCIImageIndex:
+				innerIdx, err := idx.ImageIndex(currentHash)
+				if err != nil {
+					return nil, err
+				}
+				// recursive call
+				processedIdx, err := processImageIndex(innerIdx)
+				if err != nil {
+					return nil, err
+				}
+				resultIdx = processedIdx
+				// making an assumption here that at any given point in the parent/child
+				// hierarchy, there's only a single image index entry
+				return resultIdx, nil
+			case types.DockerManifestSchema2:
+				v2format = true
+			case types.OCIManifestSchema1:
+				v2format = false
```

Comment:
Hi @jchunkins 
Question about this: we set v2format to false , because we consider that any other media types are not container images and can be ignored. Is this correct? 
ie. v2format=false can only mean this is a oci manifest

Reply:
TL;DR: v2format=false means OCI formatted media types

In **_theory_** a manifest list can contain images that are of different media types. For example, one could be a docker **_v1_** image (i.e. `application/vnd.docker.distribution.manifest.v1+json`) and another docker **_v2s2_** image (`application/vnd.docker.distribution.manifest.v2+json`), or an OCI image (i.e. `application/vnd.oci.image.manifest.v1+json`) , but in practice mixing of different types never happens. Also the `v1` image type is really rare and not really something we need to worry about, but we handle this case (see below).

The switch statement covers all of the cases we're going to care about:

- "manifest list" types
  - process recursively
- application/vnd.docker.distribution.manifest.v2+json
  - set v2format to true
- application/vnd.oci.image.manifest.v1+json
  - set v2format to false
- default
  - error out since we hit a mediatype we don't support

As you can see in the code we're using `v2format` to pick the right layer media type, and eventually we use it again to make sure the index has the right media type based on what the images are so they match.